### PR TITLE
feat: request snippets plugin

### LINKED ISSUE #6256: Would like the ability to create code snippets in various languages using swagger-ui-react
### Content & configuration

Swagger/OpenAPI definition:
[emasset.api.v1.swagger.json.txt](https://github.com/swagger-api/swagger-ui/files/4963000/emasset.api.v1.swagger.json.txt)

 Swagger-UI configuration options:
 
[swagger.js.txt](https://github.com/swagger-api/swagger-ui/files/4963011/swagger.js.txt)

### Is your feature request related to a problem?
Would like the ability to create code snippets in various languages using swagger-ui-react.

### Describe the solution you'd like
I would like to be able to use swagger-ui to create code examples in various languages.
Ideally this could be configured using extensions in the openapi document and would be displayed via swagger-ui-react.
As an example you can see [this document](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-audit-records/#api-rest-api-3-auditing-record-get) 
Screen Shot 2020-07-22 at 3 34 37 PM

Having a look at this functionality it appears to have been built as follows
* [readme.com](https://readme.com/) leverages [api extensions](https://docs.readme.com/guides/docs/openapi-extensions)
* Specifically you can set the following in your swagger.json
```
  "x-samples-enabled": true,
  "x-samples-languages": ["curl", "node", "ruby", "javascript", "python","go","c"],
```
* This then leverages [httpsnippet](https://www.npmjs.com/package/httpsnippet)

### Describe alternatives you've considered
Currently we're using 
* [docusarus](https://docusaurus.io/)
* And creating a page using [swagger-ui-react](https://www.npmjs.com/package/swagger-ui-react)
* Also have done some initial investigation into [api](https://www.npmjs.com/package/api)

### Additional context
I'd like to thank and send encouragement to the team for building out this highly useful software.
It looks like a number of teams are using this in conjunction with [docusarus](https://docusaurus.io/)
Some of which are running into [this build issue](https://github.com/swagger-api/swagger-ui/issues/6136)

### LINKED ISSUE #6908: Swagger alters request body before sending to API
### Environment
 - OS: Windows 7 | Mac OS 11.1
 - Browser: Chrome/Firefox | Safari
 - Version: 88.0.4324.104/85.0 | 14.0.2
 - Method of installation: Nuget
 - Swagger-UI version: 5.6.3

### Describe the bug
Post Request bodies that contain duplicate property names in a JSON body are altered by Swagger-UI before being posted the target API.

### To reproduce
Make a post request that has duplicate properties such as the following:
`
{
    "name": "Recipient list ToCCBBcc will it fail",
    "recipientType": "To",
    "recipients": [1, 2],
    "recipientType": "Cc",
    "recipients": [3],
    "recipientType": "Bcc",
    "recipients": [4]
}
`

Swagger will actually POST the body to the API as:
`
{
  "name": "Recipient list ToCCBBcc will it fail",
  "recipientType": "Bcc",
  "recipients": [4]
}
`

Swagger changes the body without warning or message. 
Swagger also updates the curl example command using the same body that is submitted.

### Expected behavior
Either of the following option although #1 is preferred.

1) Swagger submits the body as entered to the API and displays the same body in the curl command example.
2) Swagger refuses to submit the body due to duplicate properties in JSON object body.

### Screenshots
* Chrome v88.0.4324.102 on Windows 7
2021-02-04-Chrome-Swagger

* Firefox v85.0 on Windows 7
2021-02-04-Firefox-Swagger

* Safari v14.02 on MacOS 11.1
2021-02-04-Safari-Swagger PNG

### Additional context or thoughts
* Once we encountered this issue we returned to several previous API projects and found the same behavior across all our APIs despite different version of swagger and variations in configuration.
* Although only tested on POST I suspect the same behavior on any call with a body.
* Anything that changes the request as entered by a user should be avoided and if unavoidable should be well documented.

### PR DESCRIPTION
**Pull request by mathis-m**
### Description

Swagger ui internally stores the body some times as string e.g. when application/json.
If this is the case and the stored string is parsable to a JSON object or array it will use the parsed version for the request.
This sounds little bit weird so I removed the util and conditional path to parse the input, before creating the request.
This is to ensure that the user entered request body is NEVER overridden with the parsed value of itself.
 
I have refactored `curlify.js` to a snippets plugin (#6256):

https://user-images.githubusercontent.com/11584315/109593337-098f9f80-7b11-11eb-8080-48f4d7aac101.mp4

#### TODO:
add tests for snippet plugin clean curlify leftovers.

#### Default Configuration for plugin:
```js
requestSnippets: {
  generators: {
	"curl_bash": {
	  title: "cURL (bash)",
	  syntax: "bash"
	},
	"curl_powershell": {
	  title: "cURL (PowerShell)",
	  syntax: "powershell"
	},
	"curl_cmd": {
	  title: "cURL (CMD)",
	  syntax: "bash"
	},
	"node_native": {
	  title: "Node.js (Native)",
	  syntax: "javascript"
	},
  },
  defaultExpanded: true,
  languagesMask: null, // e.g. only show curl bash = ["curl_bash"]
},
```
### Motivation and Context

Fixes #6908
Fixes #6256
### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by mathis-m** [comment]
@tim-lai Is the curlify refactor something you like? I think it would be best to cap the height at some point and make it expandable.

**Comment by mathis-m** [comment]
Going to split this into raw request body PR and a PR for the code gen. Probably this is going to be a own plugin.

**Comment by mathis-m** [comment]
failing due to webpack `maxAssetSize: 1072000`

**Comment by tim-lai** [comment]
@mathis-m I think this is a great feature. Some comments:

- Fyi, regarding (not) using a parsed version, I think this PR helps, though I believe the source data is coming from `mutatedRequest` state via swagger-client. I also believe that the existing Curl component is the only user of the `mutatedRequest` state. I'm not sure we should be changing the interface between swagger-client and `mutatedRequest`, so my opinion is that this PR is fine for this concern.

- I don't know enough about the actual various environment escaping differences. Otherwise it looks good that you've covered for their differences.

- bundlesize... I'm curious, how much it is due to adding more syntax highlighting languages, vs. new plugin?

**Comment by mathis-m** [comment]
@tim-lai 
image
**Highlight**
javascript - 14.04 KB
powershell - 7.93
request-snippets plugin - 18.4 KB

**Comment by mathis-m** [comment]
the old curlify is not cleaned. So bundle size will be reduced by  5.97 KB

**Comment by mathis-m** [comment]
@tim-lai what do you think of creating group by eg. ( cURL > bash | powershell | cmd, node.js > native | node-fetch, powershell > Invoke-WebRequest | Invoke-RestMethod | HttpClient from .Net Namespace)

**Comment by tim-lai** [comment]
> @tim-lai what do you think of creating group by eg. ( cURL > bash | powershell | cmd, node.js > native | node-fetch, powershell > Invoke-WebRequest | Invoke-RestMethod | HttpClient from .Net Namespace)

Not sure I understand the question... can you clarify/rephrase the question? Thanks!

**Comment by tim-lai** [comment]
> the old curlify is not cleaned. So bundle size will be reduced by 5.97 KB + x for jsx.

Ok, thanks. In a separate PR, I am inclined to bump the bundlesize limit by a small amount.

**Comment by mathis-m** [comment]
> > @tim-lai what do you think of creating group by eg. ( cURL > bash | powershell | cmd, node.js > native | node-fetch, powershell > Invoke-WebRequest | Invoke-RestMethod | HttpClient from .Net Namespace)
> 
> Not sure I understand the question... can you clarify/rephrase the question? Thanks!

I thought it might look better if we add the opportunity to group several request snippets under a drop-down menu.
Eg. :

cURL
  - bash
  - powershell
  - cmd

Node
  - native
  - node-fetch
...

**Comment by tim-lai** [comment]
> I thought it might look better if we add the opportunity to group several request snippets under a drop-down menu.

I think adding dropdowns at this time adds additional ux considerations. So for now, I suggest keeping a flat tab menu.

**Comment by mathis-m** [comment]
@tim-lai 
I have:
- cleaned the old leftovers.
- redirected curlify tests to new fn
Should we add tests for new snippet generator functions(curl powershell, curl cmd, nodejs native)?

**Comment by tim-lai** [comment]
@mathis-m Two new comments:

1. Yes, should add tests for new snippet generator
2. Removing the `Curl` component could be classified as a breaking change, which we don't want to do.

File: docs/usage/configuration.md [hunk 1]
```diff
@@ -65,6 +65,8 @@ Parameter name | Docker variable | Description
...
+<a name="tryItOutEnabled"></a>`requestSnippets` | _Unavailable_ | `Object`. This is the default configuration section for the the requestSnippets plugin.<br>requestSnippets: {<br>&nbsp;&nbsp;generators: {<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_bash": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (bash)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "bash"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_powershell": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (PowerShell)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "powershell"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_cmd": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (CMD)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "bash"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"node_native": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "Node.js (Native)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "javascript"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;},<br>&nbsp;&nbsp;defaultExpanded: true,<br>&nbsp;&nbsp;languagesMask: null, // e.g. only show curl bash = \["curl_bash"\]<br>},
```
Comment (tim-lai) [inline docs/usage/configuration.md:68]:
```suggestion
<a name="requestSnippets"></a>`requestSnippets` | _Unavailable_ | `Object`. This is the default configuration section for the the requestSnippets plugin.<br>requestSnippets: {<br>&nbsp;&nbsp;generators: {<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_bash": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (bash)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "bash"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_powershell": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (PowerShell)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "powershell"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"curl_cmd": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "cURL (CMD)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "bash"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;&nbsp;&nbsp;"node_native": {<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;title: "Node.js (Native)",<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;syntax: "javascript"<br>&nbsp;&nbsp;&nbsp;&nbsp;},<br>&nbsp;&nbsp;},<br>&nbsp;&nbsp;defaultExpanded: true,<br>&nbsp;&nbsp;languagesMask: null, // e.g. only show curl bash = \["curl_bash"\]<br>},
```

**Comment by mathis-m** [comment]
@tim-lai would this mean that the plugin can not be enabled by default and the old curl component would be used instead on default?

**Comment by tim-lai** [comment]
> @tim-lai would this mean that the plugin can not be enabled by default and the old curl component would be used instead on default? Or should I try to integrate the curl into the request snippets plugin, but since I have splitter curl it might not make sense to do so. Another option would be to provide the old curl as extra tab.

Right, the old `Curl` component should be used by default, but the component itself can be changed to reflect updates. Another possibility is to leave old `Curl` as-is. However, I would prefer to maintain only 1 set of logic that is shared between `Curl` and `requestSnippets`, though either way, should mark the old `Curl` component for deprecation in the next major release.

**Comment by tim-lai** [comment]
@mathis-m PR merged! Thanks for the new feature!

**Comment by andrei572** [comment]
From documentation perspective, would it be possible to post a complete working example configuration or step by step to demo how to enable requestSnippets? I am not able to make it work with SwaggerUI, see below. What I am doing wrong?

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta
      name="description"
      content="SwaggerUI"
    />
    <title>SwaggerUI</title>
    <link rel="stylesheet" href="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui.css" />
  </head>
  <body>
  <div id="swagger-ui"></div>
  <script src="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui-bundle.js" crossorigin></script>
  <script src="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui-standalone-preset.js" crossorigin></script>
  <script>
    window.onload = () => {
      window.ui = SwaggerUIBundle({
        url: './openapi.yaml',
        dom_id: '#swagger-ui',
        presets: [
          SwaggerUIBundle.presets.apis,
          SwaggerUIStandalonePreset
        ],
        plugins: [
           SwaggerUIBundle.plugins.requestSnippets
        ],
        layout: "StandaloneLayout",
        supportedSubmitMethods: [],
        displayOperationId: true,
        requestSnippetsEnabled: true,
  requestSnippets: {
    generators: {
      "curl_bash": {
        title: "cURL (bash)",
        syntax: "bash"
      },
      "curl_powershell": {
        title: "cURL (PowerShell)",
        syntax: "powershell"
      },
      "curl_cmd": {
        title: "cURL (CMD)",
        syntax: "bash"
      },
      "node_native": {
        title: "Node.js (Native)",
        syntax: "javascript"
      }
    },
    defaultExpanded: true,
    languagesMask: null // e.g. only show curl bash = ["curl_bash"]
  }
      });
    };
  </script>
  </body>
</html>
```

**Comment by mathis-m** [comment]
> From documentation perspective, would it be possible to post a complete working example configuration or step by step recipe to demo how to enable `requestSnippets`? I am not able to make it work with SwaggerUI, "Snippets" section is not showing up. See my **index.html** below. What am I doing wrong?
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
>   <head>
>     <meta charset="utf-8" />
>     <meta name="viewport" content="width=device-width, initial-scale=1" />
>     <meta
>       name="description"
>       content="SwaggerUI"
>     />
>     <title>SwaggerUI</title>
>     <link rel="stylesheet" href="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui.css" />
>   </head>
>   <body>
>   <div id="swagger-ui"></div>
>   <script src="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui-bundle.js" crossorigin></script>
>   <script src="https://unpkg.com/swagger-ui-dist@4.14.0/swagger-ui-standalone-preset.js" crossorigin></script>
>   <script>
>     window.onload = () => {
>       window.ui = SwaggerUIBundle({
>         url: './openapi.yaml',
>         dom_id: '#swagger-ui',
>         presets: [
>           SwaggerUIBundle.presets.apis,
>           SwaggerUIStandalonePreset
>         ],
>         plugins: [
>            SwaggerUIBundle.plugins.requestSnippets
>         ],
>         layout: "StandaloneLayout",
>         supportedSubmitMethods: [],
>         displayOperationId: true,
>         requestSnippetsEnabled: true,
>   requestSnippets: {
>     generators: {
>       "curl_bash": {
>         title: "cURL (bash)",
>         syntax: "bash"
>       },
>       "curl_powershell": {
>         title: "cURL (PowerShell)",
>         syntax: "powershell"
>       },
>       "curl_cmd": {
>         title: "cURL (CMD)",
>         syntax: "bash"
>       },
>       "node_native": {
>         title: "Node.js (Native)",
>         syntax: "javascript"
>       }
>     },
>     defaultExpanded: true,
>     languagesMask: null // e.g. only show curl bash = ["curl_bash"]
>   }
>       });
>     };
>   </script>
>   </body>
> </html>
> ```

Have a look at [swagger ui docs](https://swagger.io/docs/open-source-tools/swagger-ui/customization/plug-points/)

Tested is working.

**Pull request merged**